src/main/java/com/redhat/devtools/intellij/lsp4mp4ij/psi/core/PropertiesManagerForJava.java

```diff
@@ -370,6 +354,90 @@ public Hover hover(MicroProfileJavaHoverParams params, IPsiUtils utils) {
+     */
+    public JavaCursorContextResult javaCursorContext(MicroProfileJavaCompletionParams params, IPsiUtils utils) {
+        String uri = params.getUri();
+        PsiFile typeRoot = resolveTypeRoot(uri, utils);
+        if (!(typeRoot instanceof PsiJavaFile)) {
+            return new JavaCursorContextResult(JavaCursorContextKind.IN_EMPTY_FILE, "");
+        }
+        Document document = PsiDocumentManager.getInstance(typeRoot.getProject()).getDocument(typeRoot);
+        if (document == null) {
+            return new JavaCursorContextResult(JavaCursorContextKind.IN_EMPTY_FILE, "");
+        }
+        Position completionPosition = params.getPosition();
+        int completionOffset = utils.toOffset(document, completionPosition.getLine(), completionPosition.getCharacter());
+
+        //CompilationUnit ast = ASTResolving.createQuickFixAST((ICompilationUnit) typeRoot);
+
+        JavaCursorContextKind kind = getJavaCursorContextKind((PsiJavaFile)typeRoot, completionOffset);
+        String prefix = getJavaCursorPrefix(document, completionOffset);
+
+        return new JavaCursorContextResult(kind, prefix);
+    }
+
+    private static @NotNull JavaCursorContextKind getJavaCursorContextKind(PsiJavaFile javaFile, int completionOffset) {
+        if (javaFile.getClasses().length == 0) {
+            return JavaCursorContextKind.IN_EMPTY_FILE;
+        }
+
+        PsiElement element = javaFile.findElementAt(completionOffset);
+        PsiElement parent = PsiTreeUtil.getParentOfType(element, PsiModifierListOwner.class);
+
+        if (parent == null) {
+            // We are likely before or after the class declaration
+            PsiElement firstClass = javaFile.getClasses()[0];
+
+            if (completionOffset <= firstClass.getTextOffset()) {
+                return JavaCursorContextKind.BEFORE_CLASS;
+            }
+
+            return JavaCursorContextKind.NONE;
+        }
+
+        if (parent instanceof PsiClass) {
+            PsiClass psiClass = (PsiClass) parent;
+            if (psiClass.isAnnotationType()) {
+                return JavaCursorContextKind.IN_CLASS_ANNOTATIONS;
+            }
+            return JavaCursorContextKind.IN_CLASS;
+        }
+
+        if (parent instanceof PsiMethod) {
+            PsiMethod psiMethod = (PsiMethod) parent;
+            if (psiMethod.getReturnType() != null &&  psiMethod.getReturnTypeElement() != null && completionOffset <= psiMethod.getReturnTypeElement().getTextOffset()) {
+                return JavaCursorContextKind.BEFORE_METHOD;
+            }
+            return JavaCursorContextKind.IN_METHOD_ANNOTATIONS;
+        }
+
+        if (parent instanceof PsiField) {
+            PsiField psiField = (PsiField) parent;
+            @Nullable PsiTypeElement fieldType = psiField.getTypeElement();
+            if (fieldType != null && completionOffset <= fieldType.getTextOffset()) {
+                return JavaCursorContextKind.BEFORE_FIELD;
```

Comment:
I'm not sure this will work according to the definition in JavaCursorContextKind line 87. Before_Field means the cursor is in whitespace before a field. E.g. `int a;  |  int b;` In this code the parent of the whitespace is the containing class rather than a field (`b` in this example). The same concept applies to methods also.

Reply:
the whole JavaCursor thing is mostly wrong for now, I'm working on it. I used chatGPT to quickly generate the initial PSI-based implem.

Reply:
@turkeylurkey can you check how the latest commit works for you?